### PR TITLE
enhancement(vrl)!: implement template strings in VRL

### LINKED ISSUE #7117: Add ability to easily format strings in VRL
Requested by user in discord: https://discord.com/channels/742820443487993987/764187584452493323/831828809358770226

Currently users can use `+` to concat strings, but this is a bit unwieldily for creating formatted strings. It would be nice to support some sort of sprintf-style function.

### PR DESCRIPTION
**Pull request by StephenWakely**
Closes #7117
Ref #8467 

This implements template strings in VRL.

This initial version is the simplest version allowing for future expansion should the need arise. A string is templated by delimiting with `{{..}}`. Currently only variables that resolve to strings are supported. Templates can be escaped using `/{{../}}`.

```coffee
thing = "balloon"
string = "4 red {{ thing }}s"
```

### CODE REVIEW

File: lib/vrl/compiler/src/expression/literal.rs [hunk 1]
```diff
@@ -51,7 +51,10 @@ impl TryFrom<Node<ast::Literal>> for Literal {
         let (span, lit) = node.take();
 
         let literal = match lit {
-            String(v) => Literal::String(Bytes::from(v)),
+            // Template strings have to be rewritten and compiled as an
+            // expression. Hence this variant should never be invoked.
+            String(_) => unreachable!(),
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/expression/literal.rs:56]: This seems like an undesirable design. If template strings are not literals (they can be, if they only reference other literals), then they shouldn't be exposed as a literal expression.

Put another way, can we come up with a design that doesn't require this `unreachable`?

File: lib/vrl/compiler/src/expression/object.rs [hunk 2]
```diff
@@ -1,25 +1,26 @@
...
     state::{ExternalEnv, LocalEnv},
+    value::VrlValueConvert,
     vm::OpCode,
     Context, Expression, TypeDef, Value,
 };
 
 #[derive(Debug, Clone, PartialEq)]
 pub struct Object {
-    inner: BTreeMap<String, Expr>,
+    inner: Vec<(Expr, Expr)>,
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/expression/object.rs:13]: This is something we should have discussed in the RFC.

I believe that we should not (yet, at least) support template strings in object literals.

That is, this is allowed:

```coffee
foo = "foo {{bar}}"
```

This is not:

```coffee
foo = { "foo {{bar}}": true }
```

There might come a time when we do want to allow that, but I don't think now is the time. It would probably happen if/when we allow `."foo {{bar}}" = true` as well.
```

**Review by JeanMertz — CHANGES_REQUESTED** [review]
I didn't review the entire PR yet, as I'd like us to back out of the object key changes, which likely simplifies the PR a fair bit.

File: lib/vrl/compiler/src/expression/object.rs (hunk 2 above)
Comment (jszwedko) [reply lib/vrl/compiler/src/expression/object.rs:13]: I'm on the fence about this. To me the object key there is just a regular old string and so should support the interpolation. I should be able to use any expression that returns a string there. I see that currently doesn't work though (e.g. `foo = { upcase("bar"): 123 }` is invalid) though so I can also see the case to disallow it for consistency.

Comment (jszwedko) [reply lib/vrl/compiler/src/expression/object.rs:13]: Yeah, agreed 👍

File: lib/vrl/compiler/src/expression/literal.rs (hunk 1 above)
Comment (StephenWakely) [reply lib/vrl/compiler/src/expression/literal.rs]: I have removed the  `TryFrom` implementation, and have moved that functionality into the  `compiler::compile_literal` function. 

Now we can compile the string and have all the context required to do so.

Have also updated it to just return a `Literal::String` rather than rewrite if the template is just a single literal string.

File: lib/vrl/compiler/src/expression/object.rs (hunk 2 above)
Comment (StephenWakely) [reply lib/vrl/compiler/src/expression/object.rs]: Have reverted all these changes. Object keys are now parsed differently in `parser.lalrpop` so they are essentially treated as escaped raw strings.

File: lib/vrl/parser/src/template_string.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+impl TemplateString {
+    /// Rewrites the ast for the template string to be a series of string concatenations
+    pub fn rewrite(&self) -> Expr {
+        self.0
+            .iter()
+            .map(|node| -> Expr {
+                match node {
+                    StringSegment::Literal(s) => {
+                        Expr::Literal(Node::new(diagnostic::Span::default(), RawString(s.clone())))
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/vrl/parser/src/template_string.rs:17]: Is there a reason why we don't store the span in this variant, similar to the `Template` variant?

File: lib/vrl/parser/src/template_string.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+#[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Debug, Hash)]
+pub struct TemplateString(pub Vec<StringSegment>);
+
+impl TemplateString {
+    /// Rewrites the ast for the template string to be a series of string concatenations
+    pub fn rewrite(&self) -> Expr {
+        self.0
+            .iter()
+            .map(|node| -> Expr {
+                match node {
```
Comment (JeanMertz) [inline lib/vrl/parser/src/template_string.rs:15]: nit

```suggestion
            .map(|node| match node {
```

You'll also need to remove a single `}` at the end of this block.

File: lib/vrl/parser/src/template_string.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                    StringSegment::Literal(s) => {
+                        Expr::Literal(Node::new(diagnostic::Span::default(), RawString(s.clone())))
+                    }
+                    StringSegment::Template(s, span) => {
+                        Expr::Variable(Node::new(*span, Ident::new(s)))
+                    }
```
Comment (JeanMertz) [inline lib/vrl/parser/src/template_string.rs:21]: It might be worth adding a `From<StringSegment> for Exp {}` impl, so this while `map` can be `map(Into::into)`

File: lib/vrl/parser/src/template_string.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+use std::fmt;
+
+use crate::ast::{Expr, Ident, Literal::RawString, Node, Op, Opcode};
+
+#[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Debug, Hash)]
+pub struct TemplateString(pub Vec<StringSegment>);
+
+impl TemplateString {
+    /// Rewrites the ast for the template string to be a series of string concatenations
+    pub fn rewrite(&self) -> Expr {
```
Comment (JeanMertz) [inline lib/vrl/parser/src/template_string.rs:11]: I was going to suggest using a similar language as discussed [here](https://rustc-dev-guide.rust-lang.org/lowering.html):

```suggestion
    pub fn lower(&self) -> Expr {
```

But we're rewriting one AST node to another, so I guess that's not _exactly_ apples to apples.

Let's go with `rewrite_to_concatenated_strings` instead, to make it more clear what's going on at the call-site of this function.

File: lib/vrl/parser/src/template_string.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    /// If the template string is just a single literal string return that string
+    /// as we can just represent it in the ast as a single literal, otherwise return
+    /// None as we will need to rewrite it into an expression.
+    pub fn literal_string(&self) -> Option<String> {
+        match self.0.as_slice() {
+            [StringSegment::Literal(s)] => Some(s.clone()),
+            _ => None,
+        }
+    }
+}
```
Comment (JeanMertz) [inline lib/vrl/parser/src/template_string.rs:51]: A few notes w.r.t. conventions:

- I would suggest leaving the `clone` out of here, and leaving that as a decision of the caller.
- When doing so, renaming this to `as_literal_string` fits better, as it's a pattern promoted by Rust itself, and one we also heavily use in the `Value` type.

At the call-site, it'd become `template.as_literal_string().cloned()`.

File: lib/vrl/parser/src/template_string.rs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+pub enum StringSegment {
+    Literal(String),
+    Template(String, Span),
+}
+
+impl fmt::Display for StringSegment {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            StringSegment::Literal(s) => write!(f, "{}", s),
+            StringSegment::Template(s, _) => write!(f, "{}", s),
```
Comment (JeanMertz) [inline lib/vrl/parser/src/template_string.rs:73]: I assume we want to write `{{foo}}` instead of just `foo` in this case, as that would mean it's impossible to distinguish between an actual literal string with the value `foo`, and a variable with the same name.

File: website/content/en/highlights/2022-05-03-0-22-0-upgrade-guide.md [hunk 9]
```diff
@@ -63,3 +64,40 @@ project_id = "vector-123456"
...
+
+It should be noted that currently the placeholder must contain a simple
+variable name and that variable must resolve to a string.
+
+This will not work:
+
+```coffee
+stars = 42
+sky = "There are {{ stars }} in the sky."
+```
```
Comment (JeanMertz) [inline website/content/en/highlights/2022-05-03-0-22-0-upgrade-guide.md:90]: It's always good to offer solutions as part of the example after saying something will not work, my suggestion:


> This will not work:
> 
> 
> ```coffee
> stars = 42
> sky = "There are {{ stars }} in the sky."
> ```
> 
> 
> Instead, do this:
>
> ```coffee
> stars = to_string!(42)
> sky = "There are {{ stars }} in the sky."
> ```

File: website/content/en/highlights/2022-05-03-0-22-0-upgrade-guide.md [hunk 10]
```diff
@@ -63,3 +64,40 @@ project_id = "vector-123456"
...
+
+Also paths are currently not supported, so this will not work:
+
+```coffee
+message = "The message is {{ .message }}."
+```
```
Comment (JeanMertz) [inline website/content/en/highlights/2022-05-03-0-22-0-upgrade-guide.md:96]: Similarly here:


> Also paths are currently not supported, so this will not work:
> 
> ```coffee
> message = "The message is {{ .message }}."
> ```
> 
> Assign the field to a variable, first:
>
> ```coffee
> message = .message
> message = "The message is {{ message }}."
> ```

File: website/content/en/highlights/2022-05-03-0-22-0-upgrade-guide.md [hunk 11]
```diff
@@ -63,3 +64,40 @@ project_id = "vector-123456"
...
+
+If you wish to insert `{{` into the string, you can escape using `\{{`. You
```
Comment (JeanMertz) [inline website/content/en/highlights/2022-05-03-0-22-0-upgrade-guide.md:98]:
```suggestion
If you wish to insert `{{` into the string, you can escape using `\{{` and `\}}`. You
```

File: lib/vrl/parser/src/parser.lalrpop [hunk 12]
```diff
@@ -461,18 +463,21 @@ Array: Array = {
...
 Object: Object = {
     "{" NonterminalNewline* "}" => Object(BTreeMap::default()),
-    "{" NonterminalNewline* <CommaMultiline<(<Sp<String>> ":" <ArithmeticExpr>)>> "}" => {
+    "{" NonterminalNewline* <CommaMultiline<(<Sp<ObjectKey>> ":" <ArithmeticExpr>)>> "}" => {
         let object =<>.into_iter().collect::<BTreeMap<_, _>>();
         Object(object)
     },
 };
 
+ObjectKey: String = "string literal" => <>.unescape();
```
Comment (JeanMertz) [inline lib/vrl/parser/src/parser.lalrpop:472]:
```suggestion
#[inline]
ObjectKey: String = "string literal" => <>.unescape();
```

File: lib/vrl/parser/src/lex.rs [hunk 13]
```diff
@@ -299,55 +302,55 @@ pub enum Token<S> {
 
 impl<S> Token<S> {
     pub(crate) fn map<R>(self, f: impl Fn(S) -> R) -> Token<R> {
-        use self::Token::*;
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:302]: Why this change? It seems unrelated, and personally, I dislike the extra noise it adds.

File: lib/vrl/parser/src/lex.rs [hunk 14]
```diff
@@ -436,17 +440,84 @@ impl<'input> Token<&'input str> {
...
+pub struct RawStringLiteral<S>(pub S);
 
 impl StringLiteral<&str> {
-    pub fn unescape(&self) -> String {
-        match self {
-            StringLiteral::Escaped(s) => unescape_string_literal(s),
-            StringLiteral::Raw(s) => s.to_string(),
+    /// Takes the string and splits it into segments of literals and templates.
+    /// A templated section is delimited by `{{..}}`. ``{{` can be escaped using
+    /// `/{{.../}}`.
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:451]:
```suggestion
    /// `\{{...\}}`.
```

File: lib/vrl/parser/src/lex.rs [hunk 15]
```diff
@@ -436,17 +440,84 @@ impl<'input> Token<&'input str> {
...
+                        ));
+                    }
+                    template = false;
+                    pos += 2;
+                }
+                '\\' if !template
+                    && chars.get(pos + 1) == Some(&'{')
+                    && chars.get(pos + 2) == Some(&'{') =>
+                {
+                    // Handle open escape `/{{`.
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:478]:
```suggestion
                    // Handle open escape `\{{`.
```

File: lib/vrl/parser/src/lex.rs [hunk 16]
```diff
@@ -436,17 +440,84 @@ impl<'input> Token<&'input str> {
...
+                    current.push_str(r#"{{"#);
+                    pos += 3;
+                }
+                '\\' if !template
+                    && chars.get(pos + 1) == Some(&'}')
+                    && chars.get(pos + 2) == Some(&'}') =>
+                {
+                    // Handle close escape
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:486]:
```suggestion
                    // Handle close escape `\}}`
```

File: lib/vrl/parser/src/lex.rs [hunk 17]
```diff
@@ -436,17 +440,84 @@ impl<'input> Token<&'input str> {
...
+                    pos += 2;
+                }
+                chr => {
+                    current.push(chr);
+                    pos += 1;
+                }
+            }
         }
+
+        if !template && !current.is_empty() {
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:506]: Doesn't this mean an empty string is always marked as a template string?

File: lib/vrl/parser/src/lex.rs [hunk 18]
```diff
@@ -1184,7 +1257,7 @@ mod test {
     #![allow(clippy::print_stdout)] // tests
 
     use super::{StringLiteral, *};
-    use crate::lex::Token::*;
+    use crate::lex::Token;
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:1260]: Same here, re: unrelated change that I'd like to see reverted.

File: lib/vrl/parser/src/template_string.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (StephenWakely) [reply lib/vrl/parser/src/template_string.rs:17]: We could store the span. I didn't because the only potential error reporting that requires spans occurs on the templated sections.

This should be easy enough to change to better future proof things.

File: lib/vrl/parser/src/template_string.rs (hunk 5 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/template_string.rs:21]: I could, but it is a reference at this point so it would be `From<&StringSegment>` which is not quite as idiomatic..

Comment (JeanMertz) [reply lib/vrl/parser/src/template_string.rs:21]: I don't see why it isn't. It doesn't appear we need to clone anything, so that seems fine by me?

File: lib/vrl/parser/src/lex.rs (hunk 13 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/lex.rs:302]: We now have `Token::StringLiteral` and `Token::RawStringLiteral` enum variants. We also have  `StringLiteral` and `RawStringLiteral` structs.

By importing `self::Token::*` we get conflicts between these two types.

I can see that removing the `*` does add extra noise, but I fundamentally dislike wildcard imports because that noise does add extra information that I find useful, especially when there are potential name conflicts that arise from it.

If you feel strongly I can rename `StringLiteral` to `StringLiteralToken` to avoid the name conflict?

Comment (JeanMertz) [reply lib/vrl/parser/src/lex.rs:302]: I do feel strongly. Or, rather, I feel strongly that changing something that was chosen as a convention from the start of a project, needs to have good reasoning for why we want to move away from that convention.

"I fundamentally dislike X" doesn't cut it for me, I also fundamentally dislike certain stylistic choices in other parts of Vector, but I leave them as-is, as to (1) not clutter PRs, and (2) I know someone _liked_ those choices, since they picked them, so going my way, means also moving away from someone else's preference.

I dislike discussing these things, as it makes me come across as an annoying reviewer that debates minor issues, but I also feel like _not_ calling it out means it's free-for-all for people to change stylistic choices whenever they see something that they don't like, going back-and-forth in different PRs, so that's why I take the position of sticking with the choices made at the start of a project, unless there's a good reason (or consensus) to move away from those choices.

In this case, I can see why you did it, with the naming collision, but yeah, `StrLiteral` or `StringLiteralToken` would solve the issue 👍

File: lib/vrl/parser/src/lex.rs (hunk 17 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/lex.rs:506]: If we have an empty string we have no segments. This is then handled [here](https://github.com/vectordotdev/vector/blob/stephen/vrl_interpolation/lib/vrl/parser/src/template_string.rs#L34-L39) to create an empty literal.

File: lib/vrl/parser/src/template_string.rs (hunk 5 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/template_string.rs]: It has all changed now anyway as I am passing the span information through too.

File: lib/vrl/parser/src/lex.rs (hunk 13 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/lex.rs:302]: To be clear, I didn't mean I'm making the change just because I dislike wildcards, it's just the reason I dislike wildcards popped up and so I felt it was appropriate to remove it. But I'll make the change.

File: lib/vrl/parser/src/lex.rs (hunk 17 above)
Comment (JeanMertz) [reply lib/vrl/parser/src/lex.rs:509]: That's true, but it does strike me as odd that we'd mark an empty string as a templated string in this piece of the code, regardless if we solve that in a separate place. It seems to me we want to ensure it's correctly marked here, to avoid any potential confusion if someone else uses this code without rewriting it in the future.

File: lib/vrl/parser/src/lex.rs (hunk 18 above)
Comment (JeanMertz) [reply lib/vrl/parser/src/lex.rs:1266]: I noticed you tagged me for a re-review. I assume these will be reverted before the merge?

File: lib/vrl/parser/src/lex.rs (hunk 17 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/lex.rs:509]: All strings are marked as template strings at this stage. It's only when rewriting that it gets rewritten to either a literal string or a string concatenation operation.

File: lib/vrl/parser/src/lex.rs (hunk 18 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/lex.rs:1266]: On it now!

File: lib/vrl/parser/src/lex.rs (hunk 17 above)
Comment (JeanMertz) [reply lib/vrl/parser/src/lex.rs:511]: Ah yes, you are correct 👍

On first thought, that is somewhat of an odd approach, but given that it works, and rewriting it would probably require some rethinking here, I think this is fine to land as-is.

File: lib/vrl/parser/src/lex.rs [hunk 19]
```diff
@@ -1263,8 +1344,8 @@ mod test {
                                   bar""#);
 
         match lexer.next() {
-            Some(Ok((_, Token::StringLiteral(s), _))) => assert_eq!("foo bar", s.unescape()),
-            _ => panic!("Not a string literal"),
+            Some(Ok((_, StringLiteral(s), _))) => assert_eq!(TemplateString(vec![StringSegment::Literal("foo bar".to_string(), Span::new(1, 44))]), s.template(Span::new(0, 44))),
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:1347]: nit: it'd be nice to adhere to 100 char limit (I guess rustfmt lets it slide because it's inside the `assert_eq!` macro).

File: lib/vrl/parser/Cargo.toml [hunk 20]
```diff
@@ -10,6 +10,7 @@ build = "build.rs" # LALRPOP preprocessing
 diagnostic = { package = "vrl-diagnostic", path = "../diagnostic" }
 lookup = { path = "../../lookup" }
 arbitrary = { version = "1", optional = true, features = ["derive"] }
+once_cell = { version = "1.10", default-features = false }
```
Comment (JeanMertz) [inline lib/vrl/parser/Cargo.toml:13]: I can't find where we're actually using this?

**Review by JeanMertz — APPROVED** [review]
Nice! This looks about ready to go, except for the question about `once_cell`.

File: lib/vrl/parser/src/lex.rs (hunk 19 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/lex.rs:1347]: It's the `#[rustfmt::skip]` which is not needed here..

Comment (JeanMertz) [reply lib/vrl/parser/src/lex.rs:1347]: Ah, yeah. I added that to a bunch of tests to make sure the `~` marks align with the code they are marking.
